src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -0,0 +1,273 @@
+            else
+            {
+                _property_refs.Add(new PropertyRef(0, jsonInfo));
+            }
+        }
+
+        public JsonPropertyInfo GetProperty(Type classType, ReadOnlySpan<byte> propertyName, int propertyIndex)
+        {
+            ulong key = GetKey(propertyName);
+            JsonPropertyInfo info = null;
+
+            // First try sorted lookup.
+            int count = _property_refs_sorted.Count;
+            if (count != 0)
+            {
+                int iForward = propertyIndex;
+                int iBackward = propertyIndex - 1;
+                while (iForward < count || (iBackward >= 0 && iBackward < count))
+                {
+                    if (iForward < count)
+                    {
+                        if (TryIsPropertyRefEqual(_property_refs_sorted, propertyName, key, iForward, out info))
+                        {
+                            return info;
+                        }
+                        ++iForward;
+                    }
+
+                    if (iBackward >= 0)
+                    {
+                        if (TryIsPropertyRefEqual(_property_refs_sorted, propertyName, key, iBackward, out info))
+                        {
+                            return info;
+                        }
+                        --iBackward;
+                    }
+                }
+            }
+
+            // Then try fallback
+            for (int i = 0; i < _property_refs.Count; i++)
+            {
+                if (TryIsPropertyRefEqual(_property_refs, propertyName, key, i, out info))
+                {
+                    break;
+                }
+            }
+
+            if (info == null)
+            {
+                string stringPropertyName = JsonConverter.s_utf8Encoding.GetString(propertyName);
+                throw new InvalidOperationException($"todo: invalid property {stringPropertyName}");
+            }
+
+            _property_refs_sorted.Add(new PropertyRef(key, info));
+
+            return info;
+        }
+
+        private static bool TryIsPropertyRefEqual(List<PropertyRef> list, ReadOnlySpan<byte> propertyName, ulong key, int index, out JsonPropertyInfo info)
+        {
+            if (key == list[index].Key)
+            {
+                if (propertyName.Length <= PropertyNameKeyLength ||
+                    propertyName.Slice(PropertyNameKeyLength).SequenceEqual((ReadOnlySpan<byte>)list[index].Info.Key))
+                {
+                    info = list[index].Info;
+                    return true;
+                }
+            }
+
+            info = null;
+            return false;
+        }
+
+        private static ulong GetKey(ReadOnlySpan<byte> propertyName)
+        {
+            Debug.Assert(propertyName.Length > 0);
+
+            int length = propertyName.Length;
+
+            // Embed the propertyName in the first 6 bytes of the key.
+            ulong key = propertyName[0];
+            if (length > 1)
```

Comment:
Use `MemoryMarshal.Read` with fewer nested branches (20%+ faster)

```C#
private static ulong GetKey(ReadOnlySpan<byte> propertyName)
{
    Debug.Assert(propertyName.Length > 0);

    ulong key;
    int length = propertyName.Length;

    if (length > 3)
    {
        key = MemoryMarshal.Read<uint>(propertyName);
        if (length > 4)
        {
            key |= (ulong)propertyName[4] << 32;
        }
        if (length > 5)
        {
            key |= (ulong)propertyName[5] << 40;
        }
    }
    else if (length > 1)
    {
        key = MemoryMarshal.Read<ushort>(propertyName);
        if (length > 2)
        {
            key |= (ulong)propertyName[2] << 16;
        }
    }
    else
    {
        key = propertyName[0];
    }
           
    // Embed the propertyName length in the last two bytes.
    key |= (ulong)propertyName.Length << 48;
    return key;
}```

Reply:
I was hoping there was a better way to copy six bytes[] into a ulong. I'll give this a try but hoping there's something better.

Reply:
The problem is, our `MemoryMarshal.Read` APIs expect the length to be > sizeof(T), so we can't pass a smaller than 8 byte span to it and expect the result to work.